cirq-core/cirq/circuits/circuit_operation.py

```diff
@@ -185,24 +185,55 @@ def _parameter_names_(self) -> AbstractSet[str]:
             )
         }
 
-    def _decompose_(self) -> 'cirq.OP_TREE':
-        result = self.circuit.unfreeze()
-        result = result.transform_qubits(lambda q: self.qubit_map.get(q, q))
+    def mapped_circuit(self, deep: bool = False) -> 'cirq.Circuit':
+        circuit = self.circuit.unfreeze()
+        circuit = circuit.transform_qubits(lambda q: self.qubit_map.get(q, q))
         if self.repetitions < 0:
-            result = result ** -1
-        result = protocols.with_measurement_key_mapping(result, self.measurement_key_map)
-        result = protocols.resolve_parameters(result, self.param_resolver, recursive=False)
-        # repetition_ids don't need to be taken into account if the circuit has no measurements
-        # or if repetition_ids are unset.
-        if self.repetition_ids is None or not protocols.is_measurement(result):
-            return list(result.all_operations()) * abs(self.repetitions)
-        # If it's a measurement circuit with repetitions/repetition_ids, prefix the repetition_ids
-        # to measurements. Details at https://tinyurl.com/measurement-repeated-circuitop.
-        ops = []  # type: List[cirq.Operation]
-        for repetition_id in self.repetition_ids:
-            path = self.parent_path + (repetition_id,)
-            ops += protocols.with_key_path(result, path).all_operations()
-        return ops
+            circuit = circuit ** -1
+        has_measurements = protocols.is_measurement(circuit)
+        if has_measurements:
+            circuit = protocols.with_measurement_key_mapping(circuit, self.measurement_key_map)
+        circuit = protocols.resolve_parameters(circuit, self.param_resolver, recursive=False)
+        if deep:
+
+            def remap_op_fn(op: 'cirq.Operation'):
+                if isinstance(op, CircuitOperation):
+                    return op.mapped_circuit(deep=True)
+                return op
+
+            mapped_circuit = circuit.map_ops(remap_op_fn)
+            if self.repetition_ids is None:
+                return mapped_circuit
+            if not has_measurements:
+                return mapped_circuit * abs(self.repetitions)
+
+            # Path must be constructed from the top down.
+            rekeyed_circuit = circuits.Circuit(
+                protocols.with_key_path(circuit, self.parent_path + (rep,))
+                for rep in self.repetition_ids
+            )
+            return rekeyed_circuit.map_ops(remap_op_fn)
```

Comment:
In this case, we will have called `map_ops` twice, which I think we should avoid. Perhaps something like:
```python
if self.repetition_ids is None:
    return circuit.map_ops(remap_op_fn)
if not has_measurements:
    return circuit.map_ops(remap_op_fn) * abs(self.repetitions)

rekeyed_circuit = ...
return rekeyed_circuit.map_ops(remap_op_fn)
```

Reply:
Good point. Moved the mapping into the if-blocks.